### DISCUSSION TITLE
How do you make a dynamic page?

### DISCUSSION THREAD

**Discussion by drydenwilliams** [gatsbyjs/gatsby#33307] [Help] [answered]
I would like to create a dynamic page like in Next https://nextjs.org/docs/routing/dynamic-routes but it seems super hard to do in Gatsby-land. I would like to do the following:

1. When the user visits a report URL like: `/dashboard/reports/my-awesome-report-id` 
2. It should find the correct page template `Report.tsx` 
3. In `Report.tsx` it should get the `/:id` from the URL `/dashboard/reports/:id` 
4. Fetch some data using the `id` passed through in page props

I can't seem to figure out how you get this to work? Seems to be a lack of information about this is, especially if the page **isn't** in the filesystem. It's made on the fly when the user visits that page, **not** built ahead of time. 

Sorry to share this fellow Gatsby lovers but I just want to do this:  https://nextjs.org/docs/routing/dynamic-routes

--- 

**Note for Gatsby docs team**:  It would be great to improve the Gatsby docs on this topic. When searching for this there are so many terms around routing: "Routing Reference", "Client-only Routes", "Collection Routing", "File System Route API" and "Nested Routes" just on https://www.gatsbyjs.com/docs/how-to/routing/creating-routes/ with different Github examples too.

**Comment by KyleAMathews** [comment, marked as answer]
It sounds like this is client-only (not built) so use https://www.gatsbyjs.com/docs/reference/routing/file-system-route-api/#syntax-client-only-routes (which has identical syntax to Next.js)

**Reply by drydenwilliams** [reply]
Thanks @KyleAMathews you're my hero! It worked like a charm! And, actually just as easy as Next 👍

**Comment by imagineLife** [comment]
The [Collection Routes](https://www.gatsbyjs.com/docs/reference/routing/file-system-route-api#collection-routes) are great!  
**How can I leverage nesting of the routes more dynamically at each "level" of the nest?**
i.e. a nested route:
`/{MarkdownRemark.frontmatter__parentDir}/${MarkdownRemark.frontmatter__slug}`:  
Say I have two directories, `lunches` and `desserts`. Each have several markdown files with frontmatter that include matching `parentDir` and `pageSlug` to "fit" into each directory, something like...

```
lunches/
  sandwich.md
    {...fontmatter.parentDir:"lunches"}
    {...fontmatter.slug:"sandwich"}
  salad.md
    {...fontmatter.parentDir:"lunches"}
    {...fontmatter.slug:"salad"}
desserts/
  icecream.md
    {...fontmatter.parentDir:"desserts"}
    {...fontmatter.slug:"icecream"}
  cookies.md
    {...fontmatter.parentDir:"desserts"}
    {...fontmatter.slug:"cookies"}
```

I am successful at making a collection-page at the sandwich,salad,icecream, and cookie level. With those, I query leveraging params `$frontmatter__slug` and `$frontmatter__parentDir` variables. This appends expected `data` to the page function parameters :)...

but I cant figure out how to make a "t.o.c"-like dynamic page at the `lunches` and `desserts` level.  

I've tried adding `/{MarkdownRemark.frontmatter__parentDir}/index.js`, but with this approach I haven't figured out how to build dynamic pages the same way off of the url. Including a page query into the file isn't appending any `data` object to the default page export function props object in the same way I expect it to like the `/{MarkdownRemark.frontmatter__parentDir}/${MarkdownRemark.frontmatter__pageSlug}` does
